ZEPPELIN-511 REST API: Insert / Retrieve / Move / Delete operation for paragraph

HeartSaVioR opened the pull request:
### What is this PR for?

This issue is intended to fill gap between REST API and WebSocket operations.
For now we can only access notebook and paragraph to READONLY (not writing) via REST API but after this PR, we can insert / retrieve / move / delete paragraph via REST API.

### What type of PR is it?

Feature

### Todos

### Is there a relevant Jira issue?

https://issues.apache.org/jira/browse/ZEPPELIN-511

### How should this be tested?

Please follow the explanation of added REST APIs from rest-notebook.md.

### Screenshots (if appropriate)

### Questions:
* Does the licenses files need update? (No)
* Is there breaking changes for older versions? (No)
* Does this needs documentation? (Yes, I've addressed it.)

Leemoonsoo commented:
Thanks for adding really useful REST APIs!
Looks good to me!

prabhjyotsingh commented on zeppelin-server/src/test/java/org/apache/zeppelin/rest/ZeppelinRestApiTest.java:
```diff
@@ -454,6 +454,116 @@ public void testCronJobs() throws InterruptedException, IOException{
+    p.setText("text1");
+
+    Paragraph p2 = note.addParagraph();
+    p.setTitle("title2");
```
This variable should be p2

HeartSaVioR replied:
@prabhjyotsingh Good finding! Will fix.

prabhjyotsingh commented:
@HeartSaVioR few more things

I think you should do a clean up as well after every test case execution

    ZeppelinServer.notebook.removeNote(note.getId());

Also when I run this, line 538;

    PostMethod post = httpPost("/notebook/" + note.getId() + "/paragraph/" + p.getId() + "/move/" + 10, "");

it returns 

    {"status":"OK","message":""}
  
Which i think is also not appropriate. Since, this was an invalid operation (total nos of paragraph = 2). This should have either returned some error, or should have moved 1st paragraph to the last location.

prabhjyotsingh commented on zeppelin-server/src/test/java/org/apache/zeppelin/rest/ZeppelinRestApiTest.java:
```diff
@@ -454,6 +454,116 @@ public void testCronJobs() throws InterruptedException, IOException{
+    // insert to index 0
+    String jsonRequest2 = "{\"index\": 0, \"title\": \"title2\", \"text\": \"text2\"}";
+    PostMethod post2 = httpPost("/notebook/" + note.getId() + "/paragraph", jsonRequest2);
+    LOG.info("testInsertParagraph response2\n" + post.getResponseBodyAsString());
```
This should be also post2.

HeartSaVioR replied:
Thanks for finding, will fix.

HeartSaVioR commented:
@prabhjyotsingh 
Thanks for the review. Addressed your comments.
Regarding moving paragraph to invalid index, I think it could be treated as ```bad request```, so I applied it. Would it be fine for you?

HeartSaVioR commented:
Upmerged. Please take a look again, thanks!

HeartSaVioR commented:
Done with another upmerge.

Leemoonsoo commented:
Looks good to me.
Merge if there're no more discussions.

minahlee commented:
Would you mind to do rebase than merge? It will be easier to keep track who has credits on merged code.

HeartSaVioR commented:
@minahlee No problem. Rebased via creating new branch, cherry-pick, and reset.

minahlee commented:
@HeartSaVioR Thanks for quick response. LGTM

Leemoonsoo commented:
LGTM. Merging it into master
